### ISSUE TITLE
generate unique identifier for each request and pass to upsteram as a header

### ISSUE THREAD

**Issue by ahmadnassri** [Mashape/kong#1086]
being able to trace requests coming through Kong with some sort of unique identifier would be helpful for further tracking in the upstream, a suggested header name can be: `Kong-Request-ID` _(in accordance with #324)_

**Comment by thibaultCha** [comment]
This would be the easiest plugin to make. Beginners welcome.

**Comment by Tieske** [comment]
any requirements for the ID? The easy answer is uuid, but that is slow. Will something like 'host-name + WorkerPID + counter' do?

**Comment by thibaultCha** [comment]
The most important requirement would be true, guaranteed uniqueness, at least for some platforms which heavily rely on such a feature (for example tracking a money transaction from it's request id). This could be configurable. Also if we are worried about speed, we could benchmark a uuid C binding vs an FFI implementation.

**Comment by Tieske** [comment]
> The most important requirement would be true, guaranteed uniqueness,

which requires good random algorithms, which are not cheap

How about generating a uuid once per worker process and append a counter value.

**Comment by thibaultCha** [comment]
I could see that work, it is also nice because since it is a per-worker state, the counter can simply love as a non-global in the module. 
The trade off is on the user side, where storage gets a bit more complicated and querying/indexing potentially harder (pure uuid vs string). I think it's minor though. 
Kong aims at being as less opinionated as possible, so I think providing both options would be very easy and accommodate most of the use cases.

**Comment by ahmadnassri** [comment]
> Kong aims at being as less opinionated as possible

:+1: 

could be two different plugins too, user can pick what level of performance / accuracy they want by making choice of plugin to use. *(this true to any existing or future plugins)*

**Comment by opyate** [comment]
Re: UUID, I've seen [this](https://luarocks.org/modules/mashape/lua_uuid/0.2.0-2) (a C impl), [this](https://github.com/cloudflare/raven-lua/blob/e691212f07dc944339aa8b0f45e113c3f27eecd3/raven.lua#L110-L115) and [this](http://tieske.github.io/uuid/) (pure Lua), and [this](https://github.com/bungle/lua-resty-uuid/blob/master/lib/resty/uuid.lua) and [this](https://github.com/smallfish/lua-resty-string/blob/master/lib/resty/uuid.lua) (ffi).

Presuming Kong's `lua_uuid` is good enough...:

I would like to discuss how one might configure the plugin. This is to get stylistic issues out of the way before I start hacking at a POC.

### Example 1

```
{
  "config": {
    "header_name": "Kong-Request-ID",
    "generator": "uuid"
  }
}
```

1. `config.header_name`: Following @ahmadnassri 's suggestion of using `Kong-Request-ID` as the default header name, this could be a default in `schema.lua`.

1. `config.generator`: It will default to the UUID generator. Other generators can include `timestamp` and @Tieske 's suggestions of `host`, `pid`, and `counter` and combinations of these.

Now the big question: how to specify **combinations of generators**, and especially sub-options, like "I only want one UUID per worker".

### Example 2

```
{
  "config": {
    "header_name": "Kong-Request-ID",
    "generator": {
      "uuid": {
        "once-per-worker": true
      },
      "counter": {},
      "delimiter": "."
    }
  }
}
```


1. `config.generator.<generator-name>`: are the settings for each generator, e.g. `config.generator.uuid.once-per-worker` set to `true` follows @Tieske 's `"How about generating a uuid once per worker "` suggestion.

1. Example 2's generator config will generate a `Kong-Request-ID` with the value of a UUID (once per worker), followed by a dot, followed by a number as generated by `counter`, e.g. `"80098e50-ed40-11e5-a807-27bda77c52c8.1337"` for the 1337th request on said worker.

**This is where I'd especially like feedback:** `counter` is empty. Would it be better to specify `generator` as `"{uuid:per-worker}-{counter}"` and then parse it? Or do we stay away from DSLs? Other DSL tokens could be `per-server` or `value:1337` for static things.

### Example 3

```
{
  "config": {
    "header_name": "Kong-Request-ID",
    "generator": "{uuid:per-worker}.{counter}"
  }
}
```

So, to recap: I'm trying to think of this from the point of view of the administrator configuring the plugin, so I'd like feedback before putting pen to paper.

Lastly, a correlation-id works better with logging, so in the spirit of the Unix philosophy one can just suggest this plugin be used with one of the logging plugins so you can send `Kong-Request-ID` upstream AND see it in the logs, you know, for "correlation" purposes :)

**Comment by thibaultCha** [comment]
Nothing too fancy. "UUID" and "UUID-counter" are enough. Not sure about the naming for the second one but it doesn't matter much. 

As for the format, something like "uuid#counter" is probably suitable. It would be easier to parse for users than a dash like uuid's, so one can distinguish the worker's uuid from the counter easily.

**Comment by ajayk** [comment]
@ahmadnassri @thibaultCha if there is an header [Kong-Request-ID] already present,  the plugin should pass the Kong-Request-ID as is without modification , this would help in tracing api to api calls .

**Comment by opyate** [comment]
> As for the format, something like "uuid#counter" is probably suitable. It would be easier to parse for users than a dash like uuid's, so one can distinguish the worker's uuid from the counter easily.

Yes, this would be user-specified, e.g.: `curl ... --data 'config.generator=uuid;counter` for semi-colon or  `curl ... --data 'config.generator=uuid#counter` for hash.

@ajayk Noted!

**Comment by thibaultCha** [comment]
No I think you are over complicating this though...

**Comment by Tieske** [comment]
I agree, looks to complicated to me too.

Just an option to switch between a uuid-per-request or a uuid-per-worker+timestamp/counter should do.

**Comment by opyate** [comment]
Sure, with a hard-coded `#` delimiter, then.

**Comment by opyate** [comment]
PS I've submitted a WIP, but are struggling running the tests:

Does anyone have a list of LuaRocks servers I can use? Currently getting `Error: Could not satisfy dependency: lapis ~> 1.3.1-1 when running make test-plugins`

Have this at the moment:

```
λ cat ~/.luarocks/config.lua
rocks_servers={
  "http://luarocks.org/repositories/rocks";
  "http://luarocks.giga.puc-rio.br/";
  "http://rocks.moonscript.org";
}
```


And

```
λ luarocks show lapis

lapis 1.0.2-1 - A web framework for MoonScript & Lua
```

**Comment by opyate** [comment]
All, please review my latest changes. I've made a note in my tests about relying on one worker process for some things.

Once you're happy, I'll squash and PR.

**Comment by Tieske** [comment]
Code lgtm generally, some minor nit picks. But the changes to kong/tools/database_cache.lua can/should be reverted no?

**Comment by opyate** [comment]
Done, thanks. Please let me know your nit picks :)

**Comment by thibaultCha** [comment]
Can you please open a PR so the review process is easier? Thanks.

**Comment by opyate** [comment]
@thibaultCha Done: https://github.com/Mashape/kong/pull/1094

**Comment by thibaultCha** [comment]
Just manually merged the patch from #1094. Now applying the documentation. Thanks a lot!

We'll try to include this in 0.8 very soon.
